omero/developers/Server/LightAdmins.txt

```diff
@@ -0,0 +1,244 @@
+
+Since OMERO 5.4 the ``repo`` property of :ref:`OriginalFile <OMERO model
+class OriginalFile>` is mapped into the OMERO object model. Because the
+interpretation of an ``OriginalFile`` instance depends upon with which
+repository the file is associated, for security reasons the server
+greatly restricts the mutation of this property: users cannot simply
+switch a file from one repository to another.
+
+The server must allow some setting of ``repo``. It currently uses an
+indirect means of authenticating legitimately set values. Each running
+server has a secret key recorded in the `uuid` property of :ref:`Node
+<OMERO model class Node>`. This key is not available to OMERO clients,
+it is internal to the server. To set a new file's ``repo`` the
+:javadoc:`repository DAO
+<ome/services/blitz/repo/RepositoryDaoImpl.html>` prefixes the file's
+``name`` with the server's secret key. A database trigger recognizes
+this key from the ``node`` table, removes the prefix from the ``name``,
+then allows the value of ``repo`` to be set.
+
+
+Database triggers
+-----------------
+
+While :javadoc:`BasicACLVoter <ome/security/basic/BasicACLVoter.html>`
+and :javadoc:`OmeroInterceptor
+<ome/security/basic/OmeroInterceptor.html>` carry the bulk of the burden
+of enforcing restrictions on administrators, together with
+:javadoc:`AdminImpl <ome/logic/AdminImpl.html>` for the user and group
+management restrictions, the database system itself is also a key
+enforcement mechanism.
+
+The :javadoc:`update service <slice2html/omero/api/IUpdate.html>` is one
+means by which administrators may attempt to perform restricted
+operations. Hibernate's interceptors are not wholly suited to blocking
+exactly the prohibited actions so further barriers are built into the
+database that trigger upon specific data changes. The database must
+therefore have enough information to judge if an operation is permitted.
+OMERO 5.4 introduced two tables:
+
+``_roles``
+    notes the server's configured IDs for special users and groups, such
+    as ``root`` and ``system`` which are both usually ``0``; set by
+    :javadoc:`DBUserCheck <ome/services/util/DBUserCheck.html>` on
+    server startup
+
+``_current_admin_privileges``
+   notes the restrictions *not* applying to the current user on a
+   per-transaction basis; maintained by :javadoc:`OmeroInterceptor
+   <ome/security/basic/OmeroInterceptor.html>` and frequently cleared by
+   :javadoc:`LightAdminPrivilegesCleanup
+   <ome/security/basic/LightAdminPrivilegesCleanup.html>`
+
+An example database trigger would be ``user_config_delete_trigger`` on
+the ``experimenter_config`` table. This trigger raises an exception if,
+for example, an ``AdminPrivilege:Dance`` name with a ``false`` value is
+to be removed from the ``config`` of a member of the ``system`` group by
+a user who themself is restricted from dancing. This then prevents the
+restricted administrator from dancing even vicariously by lifting that
```

Comment:
Sorry, I just fathom and guess what ``vicariously`` might mean. The sentence seems understandable, the word is quit e unusual for me.